### ISSUE TITLE
Functional tests for OIDC authentication

### ISSUE THREAD

**Issue by adamcfraser** [couchbase/sync_gateway#1774]
Add functional tests for OIDC authentication, including:
- [x] Authentication via auth code flow 
- [x] Authentication using bearer ID token
- [x] Authentication using session cookie 
- [x] Authentication using session passed back in the body 
- [x] Tests for invalid ID tokens (expired, bad signature, wrong client)
  - [x] Create token with 5 second expiry, wait 10 seconds .. not currently configurable.  (waiting on https://github.com/couchbase/sync_gateway/issues/1847#issuecomment-226301468)
  - [x] Make up garbage token 
  - [x] Signature: `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9.TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ`  use garbage in last bit.  
- [x] Token refresh - auth code flow
- [x] Test for bring-your-own-token
- [x]  Config params, turn off session tokens (running into: https://github.com/couchbase/sync_gateway/issues/1858)
- [x]  Multiple providers
  - [x]  Define multiple providers (Google), make it the default, make sure the test endpoint is the provider
- [x]  NotAuthenticated option `notauthenticated Return an authorization error for this user` -- should get some sort of error back (401)
- [x] OIDC scope tests https://github.com/couchbase/sync_gateway/issues/1856
- [x] After successfully obtaining an ID token, POST to the public _session endpoint with that token (and no cookie) to get a new session for that token.  (note that session id won't be in the response, but set cookie will be on the header)

**Comment by adamcfraser** [comment]
Additional functional tests, based on items found during review w/ Pasin:

- [ ] After successfully obtaining an ID token, POST to the public _session endpoint with that token (and no cookie) to get a new session for that token.  (note that session id won't be in the response, but set cookie will be on the header)
- [ ] Define multiple providers targeting the same issuer in the config.  Get an ID token for one of the providers, and verify that you can use it to authenticate successfully (ID token authentication relies on looking up the provider based on the issuer defined in the token).  Requires fix for  https://github.com/couchbase/sync_gateway/issues/1878
- [ ] Define multiple providers targeting the same issuer in the config.  Authenticate and trigger user creation using one of the providers.  Authenticate as the same user via the other provider, and make sure it authenticates as the same user.  Requires fix for https://github.com/couchbase/sync_gateway/issues/1876

**Comment by tleyden** [comment]
> After successfully obtaining an ID token, POST to the public _session endpoint with that token (and no cookie) 

@adamcfraser just to confirm, it should be sending the token via this header right?

```
"Authorization": "Bearer <token>"
```

**Comment by tleyden** [comment]
I'm trying to test:

> After successfully obtaining an ID token, POST to the public _session endpoint with that token (and no cookie) to get a new session for that token. (note that session id won't be in the response, but set cookie will be on the header)

and it's returning a 401 response.

Here is the ngrep output:

https://gist.github.com/tleyden/c8c4a567182b47824a83cfec84e7592f

does anything jump out at you that the test is doing wrong?

The part I was least sure about is the `{"name": "test_testuser"}` JSON body.

**Comment by tleyden** [comment]
> Define multiple providers targeting the same issuer in the config. Get an ID token for one of the providers, and verify that you can use it to authenticate successfully 

I'm a bit confused on this.  

It sounds like the config should be:

```
            "providers":{
               "test1":{
                  "issuer":"http://localhost:4984/db/_oidc_testing",
                  "client_id":"sync_gateway",
                  "validation_key":"R75hfd9lasdwertwerutecw8",
                  "callback_url":"http://localhost:4984/db/_oidc_callback",
                  "register":true
               },
               "test2":{
                  "issuer":"http://localhost:4984/db/_oidc_testing",
                  "client_id":"sync_gateway",
                  "validation_key":"R75hfd9lasdwertwerutecw8",
                  "callback_url":"http://localhost:4984/db/_oidc_callback",
                  "register":true
               },
```

and then an ID token should be obtained against the `test1` provider, but I'm not sure what this means exactly:

> and verify that you can use it to authenticate successfully

**Comment by tleyden** [comment]
> After successfully obtaining an ID token, POST to the public _session endpoint with that token (and no cookie) to get a new session for that token. (note that session id won't be in the response, but set cookie will be on the header)

This passes after switching to latest master.

**Comment by tleyden** [comment]
I moved the remaining tests to a separate ticket: https://github.com/couchbase/sync_gateway/issues/1882

**Comment by tleyden** [comment]
PR https://github.com/couchbase/sync_gateway/issues/1882

**Comment by tleyden** [comment]
PR: https://github.com/couchbaselabs/mobile-testkit/pull/541
